src/common/range.rs

```diff
@@ -141,278 +146,454 @@ impl ByteRangeSpec {
+                    .split(',')
+                    .filter_map(|spec: &str| -> Option<ByteRangeSpec> {
+                        let spec = spec.parse();
+                        if spec.is_err() {
+                            invalid = true;
+                        }
+                        spec.ok()
+                    })
+                    .collect();
+                if invalid || specs.is_empty() {
+                    return Err(err());
                 }
-                Ok(Range::Bytes(ranges))
+                Ok(Self::Bytes(specs))
             }
-            (Some(unit), Some(range_str)) if unit != "" && range_str != "" => {
-                Ok(Range::Unregistered(unit.to_owned(), range_str.to_owned()))
+            (Some(unit), Some(set)) if !unit.is_empty() && !set.is_empty() => {
+                Ok(Self::Unregistered {
+                    unit: unit.to_string(),
+                    set: set.to_string(),
+                })
+            }
+            _ => Err(err()),
+        }
+    }
+}
 
-            },
-            _ => Err(::Error::Header)
+impl std::fmt::Display for ByteRangeSpec {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match *self {
+            ByteRangeSpec::FromTo(first, last) => write!(f, "{}-{}", first, last),
+            ByteRangeSpec::Last(len) => write!(f, "-{}", len),
+            ByteRangeSpec::AllFrom(first) => write!(f, "{}-", first),
         }
     }
 }
 
 impl FromStr for ByteRangeSpec {
-    type Err = ::Error;
+    type Err = InvalidRange;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut iter = s.trim().splitn(2, '-');
+        Ok(
+            match (parse_bound(iter.next())?, parse_bound(iter.next())?) {
+                (None, Some(len)) if len != 0 => ByteRangeSpec::Last(NonZeroU64::new(len).unwrap()),
+                (Some(first), None) => ByteRangeSpec::AllFrom(first),
+                (Some(first), Some(last)) if first <= last => ByteRangeSpec::FromTo(first, last),
+                _ => return Err(err()),
+            },
+        )
+    }
+}
+error_type!(InvalidRange);
 
-    fn from_str(s: &str) -> ::Result<ByteRangeSpec> {
-        let mut parts = s.splitn(2, '-');
+fn err() -> InvalidRange {
+    InvalidRange { _inner: () }
+}
 
-        match (parts.next(), parts.next()) {
-            (Some(""), Some(end)) => {
-                end.parse().or(Err(::Error::Header)).map(ByteRangeSpec::Last)
-            },
-            (Some(start), Some("")) => {
-                start.parse().or(Err(::Error::Header)).map(ByteRangeSpec::AllFrom)
-            },
-            (Some(start), Some(end)) => {
-                match (start.parse(), end.parse()) {
-                    (Ok(start), Ok(end)) if start <= end => Ok(ByteRangeSpec::FromTo(start, end)),
-                    _ => Err(::Error::Header)
-                }
-            },
-            _ => Err(::Error::Header)
+fn reduce_bound(bound: Bound<&u64>, start: bool) -> Option<u64> {
```

Comment:
nitpick: this feels like it should be two different functions instead of taking a `bool`, especially since in the `start==true` case what you really want to return is a `u64` rather than an `Option<u64>`

Reply:
yeah, I've tried out how it looks, that's better indeed.